src/js/tracing/nativeframes.ts

```diff
@@ -0,0 +1,263 @@
+import { Event, EventProcessor, Measurements } from "@sentry/types";
+import { logger, timestampInSeconds } from "@sentry/utils";
+
+import { NativeFramesResponse } from "../definitions";
+import { NATIVE } from "../wrapper";
+import { instrumentChildSpanFinish } from "./utils";
+
+interface FramesMeasurements extends Measurements {
+  frames_total: {
+    value: number;
+  };
+  frames_slow: {
+    value: number;
+  };
+  frames_frozen: {
+    value: number;
+  };
+}
+
+const MARGIN_OF_ERROR_SECONDS = 0.05;
+
+/**
+ * Instrumentation to add native slow/frozen frames measurements onto transactions.
+ */
+export class NativeFramesInstrumentation {
+  private _finishFrames: Map<
+    string,
+    { timestamp: number; nativeFrames: NativeFramesResponse | null }
+  > = new Map();
+  private _framesListeners: Map<string, () => void> = new Map();
+  private _lastSpanFinishFrames?: {
+    timestamp: number;
+    nativeFrames: NativeFramesResponse;
+  };
+
+  public constructor(
+    addGlobalEventProcessor: (e: EventProcessor) => void,
+    doesExist: () => boolean
+  ) {
+    logger.log(
+      "[ReactNativeTracing] Native frames instrumentation initialized."
+    );
+
+    addGlobalEventProcessor((event) => {
+      if (doesExist()) {
+        return this._processEvent(event);
+      }
+
+      return event;
+    });
+  }
+
+  /**
+   * To be called when a transaction is started
+   */
+  public onTransactionStart(transaction: Transaction): void {
+    void NATIVE.fetchNativeFrames().then((framesMetrics) => {
+      if (framesMetrics) {
+        transaction.setData("__startFrames", framesMetrics);
+      }
+    });
+
+    instrumentChildSpanFinish(transaction, this._onSpanFinish.bind(this));
+  }
+
+  /**
+   * To be called when a transaction is finished
+   */
+  public onTransactionFinish(transaction: Transaction): void {
+    void this._fetchFramesForTransaction(transaction);
+  }
+
+  /**
+   * Called on a span finish to fetch native frames to support transactions with trimEnd.
+   */
+  private _onSpanFinish(_: Span, endTimestamp?: number): void {
+    if (!endTimestamp) {
+      const timestamp = timestampInSeconds();
+
+      void NATIVE.fetchNativeFrames().then((nativeFrames) => {
+        if (nativeFrames) {
+          this._lastSpanFinishFrames = {
+            timestamp,
+            nativeFrames,
+          };
+        }
+      });
+    }
+  }
+
+  /**
+   * Returns the computed frames measurements and awaits for them if they are not ready yet.
+   */
+  private async _getFramesMeasurements(
+    traceId: string,
+    finalEndTimestamp: number,
+    startFrames: NativeFramesResponse
+  ): Promise<FramesMeasurements | null> {
+    if (this._finishFrames.has(traceId)) {
+      return this._prepareMeasurements(traceId, finalEndTimestamp, startFrames);
+    }
+
+    return new Promise((resolve) => {
+      const timeout = setTimeout(() => {
+        this._framesListeners.delete(traceId);
+
+        resolve(null);
+      }, 2000);
```

Comment:
Should this be configurable?

Reply:
I don't think it should, this is just a very generous arbitrary timeout to avoid leaking.